### PR TITLE
Initial saml login support

### PR DESCRIPTION
**Pull request by tymees**
This PR adds initial work for SAML support. I'm currently leaving it in Draft status while I'm conducting acceptation tests with the UU IdP. However, I'd like some initial comments from you guys on _how_ I added the support.

For development purposes I made sure the portal can run both in 'local' and 'saml' mode. Thus, if no SAML settings are provided, it will only use Django's built-in auth. If there is, and the app is running with `DEBUG = True`, it will offer both but default to SAML. (Thus, local login is only possible through the new landing view or using the normal login view directly). 
On production, it will only offer SAML auth to users but keep the normal auth enabled in the background.

The saml settings are provided through a new ``saml_settings.py`` file in ``fetc``, similar to how we load in ``ldap_settings.py``. I provided an example of the new file to use with the local dev IdP. Before merging I'm also going to add some documentation on how to actually set that up, but before I can do that I'll need to finish the general SAML documentation...

I've also added a new ``LandingView`` to replace the secondary function the original login view was serving as a landing page. Otherwise, one would be dropped into SAML in production environments before seeing the portal itself. As an added bonus, this made adding SystemMessages to the unauthenticated landing possible, so I also added that.

### CODE REVIEW

**Comment by tymees** [comment]
Small update: I changed the buttons display logic to be more tunable instead of being automagical.

Now, both links are conditionally displayed depending on new settings flags (`SHOW_DJANGO_LOGIN` and `SHOW_SAML_LOGIN`). 

I was thinking on how to actually deploy this stuff to production, and ran into a timing issue. Basically, we need to have SAML up and running already before we can even request our app to be added to the IdP. Thus, logging in would be impossible in the original situation while we were waiting on ITS.
This way, we can setup SAML while keeping the LDAP auth active (and as primary), and switching over would only be a simple settings change.

**Comment by tymees** [comment]
Needed deploy changes tracked in https://gitlab.im.hum.uu.nl/portaldev/puppet-fetc/-/merge_requests/1

**Comment by tymees** [comment]
It seems to be working now :D

However, I need to fix some things in ``cdh.federated_auth`` before this can be merged.

**Comment by tymees** [comment]
Which is now done, un-drafting this PR now.

Full reviews are now welcome! :D

**Review by miggol — APPROVED** [review]
Save for the minor things mentioned above, there are two main things I want to remark about this implementation.

#### Make `saml_settings.example.py` more verbose

This is the first file anybody trying to get this working is going to open. Everything in `cdh.federated_auth` and `create_saml_config` is super clean and well documented. But it is also quite far away from the app you're working on.

I think that on top of having all the defaults and informative comments back in fedauth, duplicating `required_attributes` back into saml_settings.py would be nice. Add a comment about the `entity_id` being the metadata url in most cases and everything gets a lot more user friendly. 

It would be nice if `saml_settings.example.py` felt like the default django `settings.py`: friendly and tutorial-like. And also immediately telling you what attributes you can expect to get from a UU IDP.

(just sharing my experience, thinking of how this would be deployed to other CDH apps)

#### SAML-aware user objects?

As far as I can read, we're doing everything dead stock on the ethics user management side of things. No custom user objects, so they just get created and populated on first login and not changed after. Existing local users get logged in to, no questions asked.

Firstly I think we could talk about which attributes we want to update at what point. Emails and names change and we might want to reflect those changes in the portal. If we drop this on prod as is there will be loads of users with information that was not updated since their first LDAP login.

Rather out of scope, but I think something like a list of always updating attributes could see some use. Same as never-update attributes. To me it all feels very up in the air right now.

Mind you, this makes me think of a hypothetical where the secretary's email would get overwritten. And other issues with solis-ID reuse.

Either way I think updating attributes deserves some thinking about. Even if the final answer is indeed to keep everything dead stock. Right now this all a little too much "trust the black magic", rather than letting me see what has been defined.

But yeah, other than these two feelings of where things could improve, it's awesome that SAML2 auth is working now and that the rest of CDH might benefit from your implementation. Great job!

File: fetc/settings.py [hunk 1]
```diff
@@ -92,6 +93,16 @@
...
+# Determines what login options are displayed on the landing page. NOTE: this
+# does not determine which login screen is actually used as default when using
+# a LoginRequiredMixin or similar.
+# Django login is also used by LDAP auth
+# SHOW_SAML_LOGIN is set to true if saml_settings.py is present and loaded
+SHOW_DJANGO_LOGIN = True
+SHOW_SAML_LOGIN = False
+# Debug option, adds a label to the buttons. Otherwise, the buttons are
+# identical
+SHOW_LOGIN_DESCRIPTORS = True
```
[the remark this answers is not in the document]
Comment (tymees) [reply fetc/settings.py:105]: Yeah good point, I'll do that

**Comment by tymees** [comment]
> #### Make `saml_settings.example.py` more verbose
> 
> This is the first file anybody trying to get this working is going to open. Everything in `cdh.federated_auth` and `create_saml_config` is super clean and well documented. But it is also quite far away from the app you're working on.

Good point, I'll add some more comments!

> It would be nice if `saml_settings.example.py` felt like the default django `settings.py`: friendly and tutorial-like. And also immediately telling you what attributes you can expect to get from a UU IDP.

Well idk all the attributes we can theoretically get from the UU IdP; however, I'll definitely add a description of the attributes we've requested for ethics.

> (just sharing my experience, thinking of how this would be deployed to other CDH apps)

Just to comment on this a bit; I am writing a more general guide for other apps, which will go into more detail. Thus, if you agree, I'd like to keep the additonal comments in the example file limited to ethics specifics and provide a link to the general docs. 

> As far as I can read, we're doing everything dead stock on the ethics user management side of things. No custom user objects, so they just get created and populated on first login and not changed after. Existing local users get logged in to, no questions asked.

The LDAP auth backend does actually update the account on every login by default. You have to explicitly disable that, which we don't.
(For example: my fetc account says 'Tyler' as firstname, even tough it was 'Ty' before and I didn't manually change it. In addition, it changed my email from my students to my uu email when I lost access to my students mail). 

> Firstly I think we could talk about which attributes we want to update at what point. Emails and names change and we might want to reflect those changes in the portal. If we drop this on prod as is there will be loads of users with information that was not updated since their first LDAP login.

That is a good conversation to have. Currently we only have names, usernames and emails, with 'faculty' being added with the move to SAML. 
I think all those attributes are 'change friendly', but I might be missing a situation in which they aren't? (And as stated above, the 'not-updated' account problem should not exist irl).

Nonetheless, having more control over which attributes are auto-updated is definitely not a bad idea for a generic auth backend, so I'll write a feature-request in DSC sometime in the next few days.

> Rather out of scope, but I think something like a list of always updating attributes could see some use. Same as never-update attributes. To me it all feels very up in the air right now.

Do you consider such functionality a blocker for actually switching over to SAML? If so, I would consider it in-scope for this PR. 

> Mind you, this makes me think of a hypothetical where the secretary's email would get overwritten. And other issues with solis-ID reuse.

When would the secretary's email be overwritten? 

Solis-ID reuse is a problem, yes. However, I don't see a reliably way to account for such cases.

**Comment by miggol** [comment]
> The LDAP auth backend does actually update the account on every login by default. You have to explicitly disable that, which we don't.

I must have been confused by the docs or editing the wrong user, I see now that they do in fact update at login. I'd say that's the better default. The secretary situation I'm worried about is that the secretary email would be updated from the NPM to a personal email.

Otherwise, I can't think of a field that is NOT change-friendly at this time. My gut is telling me that it would be good to have a "do not update this field" option before we strictly *need* it. But as long as the secretary exception doesn't happen this is not blocking for ethics.

> Solis-ID reuse is a problem, yes. However, I don't see a reliably way to account for such cases.

Yeah same here. Perhaps we could do Windows licence style fuzzy matching on all attributes and if too much changes: poof! New user! 
/s

**Comment by tymees** [comment]
>The secretary situation I'm worried about is that the secretary email would be updated from the NPM to a personal email. [..] But as long as the secretary exception doesn't happen this is not blocking for ethics.

AFAIK the email address isn't taken from an user account, but from ``EMAIL_FROM`` in settings. So I would say that problem should never occur as long as we keep it that way (and I don't see a reason to change it either way).

> Otherwise, I can't think of a field that is NOT change-friendly at this time. My gut is telling me that it would be good to have a "do not update this field" option before we strictly _need_ it. 

Okay, I'll update the PR with the other requested changes and write that FR for sometime in the future then. Thanks for taking a look at it :)

> Yeah same here. Perhaps we could do Windows licence style fuzzy matching on all attributes and if too much changes: poof! New user!
> /s

I know it's a joke, but the sad reality is that is probably the only thing that would work :')
With the same user-friendliness as well!

**Comment by tymees** [comment]
I think I got all the requested changes.

Are the comments good enough now?

**Comment by miggol** [comment]
> Are the comments good enough now?

Yes! They are excellent. Having the default attribute map in the example settings is great, the file is very educational now. I love it.

I added a small comment mentioning the difference between localhost and 127.0.0.1 that embarrassingly tripped me up during testing.

**Pull request merged**